### PR TITLE
Initial breakout of vN-maintenance branch changes

### PR DESCRIPTION
**Pull request by yadij**

### CODE REVIEW

**Comment by yadij** [comment]
see https://github.com/squid-cache/squid/pull/1268#discussion_r1097741975 for discussion behind this new policy.

Once this and PR #13 have been merged we will be in a position to re-attempt the Squid-7 series release.

File: release/mk-release-series.sh
Comment (rousskov) [inline release/mk-release-series.sh:106]: > PR description: @rousskov requires that each change to squid-cache/squid repository by automation be submitted as a separate PR.

Sigh. I do not require that, so I removed the above statement from the PR description and synced the rest of the PR title/description the best I could. Please adjust further as needed.

If you would like to propose a new policy, please do so (there is nothing wrong with that per se!), but please do not assign authorship of your proposals to me.

File: release/mk-release-series.sh
Comment (rousskov) [inline release/mk-release-series.sh:78]: I strongly recommend adding error detection/handling, at least for any major new/modified script commands like these ones, especially if the script is going to automatically create official pull requests based on those commands! "Detect/handle errors" is a basic development principle. It should be the default!

File: release/mk-release-series.sh
Comment (rousskov) [inline release/mk-release-series.sh]: Again, please detect/handle errors.

File: release/mk-release-series.sh
Comment (rousskov) [inline release/mk-release-series.sh]: To assist with human comprehension and simplify `git log` searches, please use some more descriptive commit/PR title than a bare version number.

I do not insist on any specific wording, but can suggest the following as a starting point:

```suggestion
    git commit -m "Start v${NEWVER} development"
```

or its more verbose version:

```suggestion
    git commit -m "Switch to v${NEWVER} series development"
```

The title should be interpreted in the master branch context where the PR will be merged, of course.

File: release/mk-release-series.sh
Comment (yadij) [reply release/mk-release-series.sh:78]: Which type of errors? from which process? with what handling?

FYI;
 * this script is run by a human who is watching the stderr / stdout display and looking for errors.
 * `origin` is the squidadm/squid repository. Intentionally different from the squid-cache/squid repository so that automation errors can be reverted with ease.
 * likewise the `$SQUID_VCS_PATH/squid-$NEWVER` and `$SQUID_WWW_PATH/` are staging locations which are published at a later time only if not manually reverted.

File: release/mk-release-series.sh
Comment (yadij) [reply release/mk-release-series.sh]: This is the title schema used by all other `vN.*` releases for many years now.
The old `Branch ...` title was relevant to explain when this was just one of a bunch of maintenance updates fixing code bugs automatically - to be squashed on merge. Now that is it going to be its own standalone merge it should be consistent with other releases.

To be clear and specific;

 * this is the commit which formally initiates v7.0.0-* snapshot / daily Release's. If it were any other number than `.0.0` it would also have a release packaging tag `SQUID_N_X_Y` assigned, package signing, etc.

 * commits to `master` branch between the execution of `mk-release-series.sh` script and the merge of this `N.0.0` PR **will not** be part of Squid-(N-1) releases. As such we must (and do) take extra care not to publish any snapshots of code with outdated version numbers.

File: release/mk-release-series.sh
Comment (yadij) [reply release/mk-release-series.sh]: I am aware that there are several types of error which could occur. But none of those can be automatically handled - they are all things like missing credentials, github permissions, or network outages. These do get reported to stderr so we can fix the cause and manually do the PR as workaround.

IMO the correct handling (such as can be automated) is already done by the `gh` tool internally. This script should ignore the lack of PR and continue with the process.

File: release/mk-release-series.sh
Comment (yadij) [reply release/mk-release-series.sh:106]: Er, what? the record is clear - you argued against and vetoed the alterations this script is/was committing to `vN-maintenance` branch automatically. So yes you did literally require this PR change by any reasonable definition of the word "requires".

What you apparently still have not grokked is that there are **many** maintenance scripts doing **many** changes to the source code. Most of them are not as visible as this one. All of them are currently committing their changes to the appropriate `vN-maintenance` branch for later (daily-ish) creation of a PR to merge the aggregate into the relevant vN series branch.

The maintenance update this PR is touching is "just" the one bit which will hopefully let us release the first v7.0.0 snapshot without you veto'ing the subsequent maintainer work.

**Comment by yadij** [comment]
Reverted the title change because each part of the new one was false or irrelevant.

"Automatically post PR with initial vN-maintenance branch changes"
 - automatic submission of a PR with this scripted code change occurs regardless of this PR 14.
 - this PR explicitly does **not** make any PR with "vN-maintenance branch changes".

The status-quo situation is that this touched script makes alterations to the source code and commits them to the `vN-maintenance` branch on staging `origin` repository. A *different* script later generates the PR for merging `origin/vN-maintenance` into `squid-cache:squid/master` branch.

This PR refactor moves ("breaks out") that alteration commit to happen on a different temporary branch called `vN-release-prep` and generates a PR for merging the temporary branch.

This PR is the first ("Initial") of a series of PRs updating maintenance automation to follow this new policy requested by @rousskov in PR #1268. Which is to say: submit each distinct automated code change in a distinct PR.

File: release/mk-release-series.sh
Comment (rousskov) [reply release/mk-release-series.sh:78]: > Which type of errors?

Any and all outcomes that `git` commands in question define as an error by exiting with a non-zero exit code.

> from which process? 

In this specific case, from the `git... && git...` shell command this change request is attached to. Personally, I would split that command into two `git` commands and handle errors from each one of them, but that improvement is not _required_ to address this particular change request:

```shell
git checkout -b v${NEWVER}-release-prep || exit $?
git push -u origin v${NEWVER}-release-prep || exit $?
```

> with what handling?

I recommend exiting with a non-zero error code (preferably the one returned by the failed command).

Detecting and handling errors is a well-established basic development practice. Two core developers should not be arguing whether to do it!

File: release/mk-release-series.sh
Comment (rousskov) [reply release/mk-release-series.sh]: > This script should ignore the lack of PR and continue with the process.

I strongly disagree. The default should be to quit and let the user handle the problem. If the user needs the process to continue, they can adjust the script accordingly.

File: release/mk-release-series.sh
Comment (rousskov) [reply release/mk-release-series.sh]: > This is the title schema used by all other vN.* releases for many years now.

Yes, the problem is not specific to the proposed change. I think all those titles should be improved, but this change request is specific to this one new title. If the new title has to be `N.0.0` for consistency with other bad titles, then so be it, but I would prefer to move forward instead.

File: release/mk-release-series.sh
Comment (rousskov) [reply release/mk-release-series.sh:106]: > PR description: @rousskov requires that each change to squid-cache/squid repository by automation be submitted as a separate PR.

> Alex: I do not require that.

> Amos: you argued against and vetoed the alterations this script is/was committing to vN-maintenance branch automatically. 

Not really. I [suggested](https://github.com/squid-cache/squid/pull/1268#discussion_r1097741975) some related improvements (e.g., which changes should, ideally, go into which PR), but I did not argue against the changes this script is committing (any polishing touches aside), and I certainly did not veto those alterations (I have never even voted against PR 1268!).

More importantly, all those sins that I have been falsely accused of do not actually imply that I require that "each change to squid-cache/squid repository by automation be submitted as a separate PR"! Some automation-created changes should be merged inside a single PR. Some should not be. Many variables affect that decision. I was not defining any new Project policy when I discussed specific improvements in a specific PR, and I certainly was not defining the policy summarized in the original version of this PR title!

> So yes you did literally require this PR change

I did not. It is impossible to logically connect anything I have _actually_ said or _actually_ did to this conclusion (not to mention to find a "literal" requirement for this PR change!). It is only possible to connect a yet another misinterpretation of what actually happened to the above conclusion. What a waste of time and nerve cells!

**Comment by rousskov** [comment]
> This PR is the first ("Initial") of a series of PRs updating maintenance automation to follow this new policy requested by @rousskov in PR #1268. Which is to say: submit each distinct automated code change in a distinct PR.

I have not requested such a policy. Please stop attributing it to me.

File: release/mk-release-series.sh
Comment (yadij) [reply release/mk-release-series.sh]: Nod. To make progress I am leaving this as-is. We can defer the topic of changing release practices to later. For now consistency is better.

File: release/mk-release-series.sh
Comment (yadij) [reply release/mk-release-series.sh:78]: Thank you. But no, an `exit` is the worst thing this particular block of code could do. All of the statements in this `if ... fi` block can be skipped by the release process and performed manually later without causing major problems. As demonstrated by the status quo code leaving the PR for an entirely separate asynchronous process to do.
 Whereas, an `exit` will leave the staging environment in a broken partially constructed state - worse than attempting and skipping this non-urgent task if it fails.

> Detecting and handling errors is a well-established basic development practice. Two core developers should not be arguing whether to do it!

Indeed. FWIW, the face-value implications of that statement are quite insulting to one or other of us. I will assume that a) you did not actually mean it, and b) we both have enough expertise to see the error handling present in this syntax.

For clarity, IMO the disagreement we have is about the specific mechanism(s) implemented to detect and handle errors in this script. We differ in both opinion + expertise and that is fine. Some of your other suggestions/requests have been useful and I have or will implement. This suggestion on this particular piece of code is a "wont do" for the reasons outlined above.

**Comment by yadij** [comment]
> > This PR is the first ("Initial") of a series of PRs updating maintenance automation to follow this new policy requested by @rousskov in PR #1268. Which is to say: submit each distinct automated code change in a distinct PR.
> 
> I have not requested such a policy. Please stop attributing it to me.

I do not understand why you apparently cannot comprehend the meaning of your own request:

> Alex:
>> The essence of my suggestion applies to mixing two unrelated automated changes just like it applies to mixing manual and automated changes, but, fortunately, we do not need to agree on the application of that principle in this PR context.
...
>> They should be improved, one way or the other. IMO, we cannot address or even justify the problems in this PR by simply saying that these problems were created by some scripts!

One thing the past decade of working with you has taught me is that when the word "should" starts being mentioned, you are going to argue until things happen your way regardless of how many years it takes. That make the request a de facto policy already.

File: release/mk-release-series.sh
Comment (rousskov) [reply release/mk-release-series.sh:78]: > Alex: Detecting and handling errors is a well-established basic development practice. Two core developers should not be arguing whether to do it!

> Amos: I will assume that a) you did not actually mean it

I did mean exactly what I said -- we should not be arguing whether code should detect/handle errors. Your "indeed" response in the second exchange means that you agree with that principle.

FWIW, if I were responding to a similar change request against my own code (with that principle in mind), then I would

* address that change request by adding a source code comment documenting that the error detection in this code are console error messages that the user is expected to notice and that the error handling in this code is to ignore (PR branch preparation and similar) errors in hope that the `gh` command creating an official PR a few lines later will either not happen or will fail as well (or something along those lines -- I would know exactly what to write because it would be my script doing exactly what I want it to do).
* respond with a GitHub comment saying that I have addressed the change request (by fixing the script to document that the lack of error detection/handling code is not an accident) and that I do not want to change the current error detection/handling approach (for some valid reasons).
 
That would allow for quick and painless progress!

> and b) we both have enough expertise to see the error handling present in this syntax.

This change request is about code that lacks error detection/handling. If you are implying that the presence of an `&&` operator addresses this change request, then I disagree: For _example_, there is no error detection/handling for the second git command. We should not be arguing about such obvious things either.

AFAICT, we disagree regarding what is appropriate/safe/acceptable/etc. in the context of this script. I have no additional arguments to make and no hope for changing your position. Fortunately, we do not need to agree for this PR to move forward. I wish we would not waste so much time and nerve cells on establishing such disagreements, but nothing I have tried so far worked, and I currently have no new ideas to try.

**Review by rousskov — COMMENTED** [review]
> One thing the past decade of working with you has taught me is that when the word "should" starts being mentioned, you are going to argue until things happen your way regardless of how many years it takes. That make the request a de facto policy already.

I would love my change request to be (polished and) adopted as an official policy! However, my change request does not match the "each change to squid-cache/squid repository by automation be submitted as a separate PR" policy or the "submit each distinct automated code change in a distinct PR" policy mis-attributed to me during this PR lifetime.

To reduce time waste, please avoid attributing your text to me. Even if we ignore the (un)ethical side of such (mis)attributions, the probability that your retelling of my words will match their actual meaning in such cases is just too small for such an approach to be practical.

P.S. Please do not misinterpret this response as an implicit agreement that the word "should" in my change requests has the alleged implications.

File: release/mk-release-series.sh
Comment (yadij) [reply release/mk-release-series.sh]: FWIW, I do not understand why you appear to be advocating that release maintainers perform secret changes to the release scripts.

This PR is a demonstration of what happens when "the user" attempts to "adjust the script accordingly" through the proper channels. Arguments with QA reviewer.

Comment (rousskov) [reply release/mk-release-series.sh]: > I do not understand why you appear to be advocating that release maintainers perform secret changes to the release scripts.

I do not advocate making releases (i.e. posting official PRs/etc.) with secret scripts. I am OK with folks adjusting any scripts while working on those scripts and/or on future releases (but not actually posting any official PRs/etc.), especially if such temporary/unofficial adjustments help improve the quality of the official script that actually makes the release at the end of that process.

> This PR is a demonstration of what happens when "the user" attempts to "adjust the script accordingly" through the proper channels. Arguments with QA reviewer.

This PR is a demonstration of what happens when new code that needs improvements is posted for review.

Comment (rousskov) [reply release/mk-release-series.sh]: In fact, a PR created by an official script that ignores errors should be treated similar to a PR created by a secret script (because the reviewer has no idea which errors the script has ignored before posting the PR -- the PR is effectively a result of an unknown code path execution)! My overall approach covers the two unwanted cases (i.e. error-ignoring official scripts creating official PRs and secret scripts creating official PRs) the same way (because their effect is similar): Avoid!

Compare those problems to a situation where a developer ran the official script that properly handles errors, that script failed, the developer made lots of "secret" (or not!) adjustments to properly rectify the situation, then rerun the official script that properly handles errors, and that script has finally succeeded, posting an official PR. In that situation, the developer still has a lot of freedom to properly address problems, but the resulting PR is indeed a product of a known official script -- a very desirable outcome!

File: release/mk-release-series.sh
Comment (kinkie) [reply release/mk-release-series.sh:78]: > > Which type of errors?
> 
> Any and all outcomes that `git` commands in question define as an error by exiting with a non-zero exit code.
> 
> > from which process?
> 
> In this specific case, from the `git... && git...` shell command this change request is attached to. Personally, I would split that command into two `git` commands and handle errors from each one of them, but that improvement is not _required_ to address this particular change request:
> 
> ```shell
> git checkout -b v${NEWVER}-release-prep || exit $?
> git push -u origin v${NEWVER}-release-prep || exit $?
> ```
> 
> > with what handling?
> 
> I recommend exiting with a non-zero error code (preferably the one returned by the failed command).

This can be more easily achieved globally by using
```
set -eu -o pipefail
```
at the beginning of the script

> Detecting and handling errors is a well-established basic development practice. Two core developers should not be arguing whether to do it!

In my understanding the discussion is not about whether errors should be detected - they are. The discussion is how to achieve that.

File: release/mk-release-series.sh
Comment (kinkie) [reply release/mk-release-series.sh]: Let's not blow this out of proportion. There are several checks that happen during the workflow, both automated and manual, including on the correctness of the emitted PR.

We might argue that checks that are manual and later in the process are wastefui, but I am highly confident about the correctness of the outcome. PRs come and go, and even if one were to be created that needs to be immediately closed because it's incorrect is not the end of the world. 
This said,
```
set -eu -o pipefail
```
at the beginning of the script will make all these problems go away, because it will make it so that any unhandled error anywhere in a pipeline will cause the script to exit with return code as the rightmost failing command in the pipeline

**Comment by kinkie** [comment]
Given that this script is "just" a support for a manual process, which happens very rarely and with ample manual and automated checks, I'm in favour of keeping it simple, it's robust enough for our current needs.
So.. let's ship it and focus on the actual release process which is what we are trying to achieve

**Review by kinkie — APPROVED** [review]
it's good enough for me

File: release/mk-release-series.sh
Comment (rousskov) [reply release/mk-release-series.sh:78]: > In my understanding the discussion is not about whether errors should be detected - they are.

This change request is attached to one place where an error is not detected (in newly added code) but should be. Frankly, I do not understand how one can look at this change request (in its natural context) and develop an understanding that errors are detected.

> This can be more easily achieved globally by using `set -eu -o pipefail`

I agree that `set -e` is much "easier", but some bash enthusiast recommend [avoiding](http://mywiki.wooledge.org/BashFAQ/105) `set -e`. This is not my area of expertise, but their arguments sound rather convincing to me. I do not recommend using that option as the solution.

File: release/mk-release-series.sh
Comment (rousskov) [reply release/mk-release-series.sh]: I hope that "Let's not blow this out of proportion" remark was in response to an earlier accusation that I am "advocating that release maintainers perform secret changes to the release scripts." rather than my basic request to detect and handle errors or my explanation that I am not guilty of the alleged advocacy.

There is nothing I can do to counter the "I am highly confident about the correctness of the outcome" and "not the end of the world" arguments in response to a basic request to detect and handle errors. I should probably just stop wasting time on reviewing this code.

File: release/mk-release-series.sh
Comment (yadij) [reply release/mk-release-series.sh:78]: Latest iteration of this script includes a lot more error catching and a "nuclearFallout" handler to purge all staged changes that can automatically be purged. Most of this was just tested with v7 branching, next use/test of the script will be in 2025 so I hope we can accept this and move on - with followup at that time when re-tested for v8.

Comment (rousskov) [reply release/mk-release-series.sh:78]: > I hope we can accept this and move on

If "this" is "this PR", then sure! I was not blocking this PR before, and I am still not blocking it now, despite my concerns about the script approach, the script quality, and the manner this PR has been handled. I have given up fighting this windmill.

**Pull request closed without merging**
